src/routes/v2/sgidAuth.ts

```diff
@@ -86,25 +90,130 @@ export class SgidAuthRouter {
-        this.sgidAuthService.retrieveSgidUserEmail(accessToken, sub)
+        this.sgidAuthService.retrieveSgidUserEmails(accessToken, sub)
       )
-      .andThen((email) => {
-        if (!email || !email.endsWith("@open.gov.sg")) {
+      .andThen((officerDetails) => {
+        if (!officerDetails || officerDetails.length === 0) {
           return errAsync(
             new AuthError(
               "Your email has not been whitelisted to use sgID login. Please use another method of login."
             )
           )
         }
-        return okAsync(email)
+        return okAsync(officerDetails)
       })
-      .andThen((email) =>
-        // Login with email
+    if (userDataRes.isErr()) {
+      const { error } = userDataRes
+      if (
+        error instanceof SgidFetchAccessTokenError ||
+        error instanceof SgidFetchUserInfoError
+      ) {
+        return res.status(500).send()
+      }
+      if (error instanceof AuthError) {
+        return res.status(401).send()
+      }
+      return res.status(500).send()
+    }
+
+    const userData = userDataRes.value
+    if (userData.length === 1) {
+      // If user only has a single email, login directly
+      await ResultAsync.fromPromise(
+        this.usersService.loginWithEmail(userData[0].email),
+        (error) => {
+          logger.error(
+            `Error while retrieving user info from database: ${error}`
+          )
+          return new DatabaseError()
+        }
+      )
+        .map((user) => {
+          const userInfo = {
+            isomerUserId: user.id,
+            email: user.email,
+          }
+          Object.assign(req.session, { userInfo })
+          return res.status(200).send({ userData })
+        })
+        .mapErr((error) => res.status(500).send())
+    } else {
+      // User has multiple emails, defer to user to select email
+      const csrfTokenExpiry = new Date()
+      // getTime allows this to work across timezones
+      csrfTokenExpiry.setTime(csrfTokenExpiry.getTime() + CSRF_TOKEN_EXPIRY_MS)
+      const cookieSettings = {
+        expires: csrfTokenExpiry,
+        httpOnly: true,
+        secure: isSecure,
+      }
+      const token = jwtUtils.signToken({ userData })
+      res.cookie(SGID_MULTIUSER_COOKIE_NAME, token, cookieSettings)
+      return res.status(200).send({ userData })
+    }
+  }
+
+  handleSgidMultiuserLogin: RequestHandler<
+    never,
+    void | ResponseErrorBody,
+    { email: string },
+    unknown,
+    never
+  > = async (req, res) => {
+    const { email } = req.body
+    const token = req.cookies[SGID_MULTIUSER_COOKIE_NAME]
+
+    const safeVerify = Result.fromThrowable(jwtUtils.verifyToken, (error) => {
+      logger.error(
+        `Error - invalid token for sgid multiuser login ${email}: ${error}`
+      )
+      return new SgidVerifyUserError()
+    })
+
+    await safeVerify(token)
+      .andThen((verifiedToken) => {
+        if (
```

Comment:
If this is going to be repeated, we can extract as a function.

Reply:
Not repeated! single use, separated to make it easier to read